drivers/flash/disk_flash.c

```diff
@@ -0,0 +1,344 @@
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (size - read_cursor);
+
+		if (write_sector < 0 || write_sector >= data->disk_total_sector_cnt) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (disk_access_read(cfg->disk_name, data->sector_buf, write_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (len_write > data->sector_buf_size - write_padding) {
+			len_write = data->sector_buf_size - write_padding;
+		}
+
+		memset(&data->sector_buf[write_padding],
+			cfg->flash_parameters.erase_value,
+			len_write);
+
+		if (disk_access_write(cfg->disk_name, data->sector_buf, write_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+		read_cursor += len_write;
+	}
+
+release_device:
+	release_device(dev);
+	return ret;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void disk_flash_pages_layout(const struct device *dev,
+				    const struct flash_pages_layout **layout, size_t *layout_size)
+{
+	const struct disk_flash_config *cfg = dev->config;
+
+	*layout = &cfg->fpl;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_parameters *disk_flash_get_parameters(const struct device *dev)
+{
+	const struct disk_flash_config *cfg = dev->config;
+
+	return &cfg->flash_parameters;
+}
+
+static const struct flash_driver_api disk_flash_api = {
+	.read = disk_flash_read,
+	.write = disk_flash_write,
+	.erase = disk_flash_erase,
+	.get_parameters = disk_flash_get_parameters,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = disk_flash_pages_layout,
+#endif
+#if defined(CONFIG_FLASH_JESD216_API)
+	.sfdp_read = NULL,
+	.read_jedec_id = NULL,
+#endif
+#if defined(CONFIG_FLASH_EX_OP_ENABLED)
+	.ex_op = NULL,
+#endif
+};
+
+static int disk_flash_init(const struct device *dev)
+{
+	int res = 0;
+	struct disk_flash_data *data = dev->data;
+	const struct disk_flash_config *cfg = dev->config;
+	size_t disk_size = 0;
+
+	k_sem_init(&data->sem, 1, 1);
+
+	res = disk_access_init(cfg->disk_name);
+	if (res != 0) {
+		LOG_ERR("init disk failed: %d", res);
+		goto init_done;
+	}
+
+	res = disk_access_ioctl(cfg->disk_name, DISK_IOCTL_GET_SECTOR_COUNT,
+				&data->disk_total_sector_cnt);
+	if (res != 0) {
+		LOG_ERR("read total sector cnt failed: %d", res);
+		goto init_done;
+	}
+
+
+	res = disk_access_ioctl(cfg->disk_name, DISK_IOCTL_GET_SECTOR_SIZE,
+				&data->disk_sector_size);
+	if (res != 0) {
+		LOG_ERR("read sector size failed: %d", res);
+		goto init_done;
+	}
+
+	if (data->disk_sector_size > data->sector_buf_size) {
+		LOG_ERR("sector size %u of disk to big for buffer %u", data->disk_sector_size,
+			(uint32_t)data->sector_buf_size);
+		res = -EINVAL;
+		goto init_done;
+	}
+
+	if (cfg->disk_offset % data->disk_sector_size != 0) {
+		LOG_WRN("offset is not aligned with disk sectors (disk sector size: %u)",
+			data->disk_sector_size);
+	}
+
+	disk_size = data->disk_sector_size * data->disk_total_sector_cnt;
```

Comment:
`size_t` may not be enough to hold the value. On 32 bit platform that is up to 4G. Is it possible that larger devices would be used?

Reply:
You are right. There is absolutely a possibility that large disks could be used since this driver supports partitioning. I adjusted this.